src/node/interfaces.cpp

```diff
@@ -942,9 +943,94 @@ class BlockTemplateImpl : public BlockTemplate
         return chainman().ProcessNewBlock(block_ptr, /*force_processing=*/true, /*min_pow_checked=*/true, /*new_block=*/nullptr);
     }
 
+    std::unique_ptr<BlockTemplate> waitNext(BlockWaitOptions options) override
+    {
+        // Delay calculating the current template fees, just in case a new block
+        // comes in before the next tick.
+        CAmount current_fees = -1;
+
+        // Alternate waiting for a new tip and checking if fees have risen.
+        // The latter check is expensive so we only run it once per second.
+        auto now{NodeClock::now()};
+        const auto deadline = now + options.timeout;
+        const MillisecondsDouble tick{1000};
```

Comment:
In commit "Add waitNext() to BlockTemplate interface" (7f6f58bad8e56c9a25ffbc0d105bed37e9958fef)

Any thoughts on whether it makes sense for this polling interval to be configurable? I could imagine a miner not really caring how expensive this check is and just wanting it to be as fast as possible?

Reply:
A miner might naively set it so low that our p2p processing is impacted, potentially getting new blocks later and interfering with transaction relay.

Also, I hope to get rid of this soon(tm) if cluster mempool gives us a more efficient method. It's nice if we don't have to drop a configuration option again.

This is already 60x faster than the long poll behavior of `getblocktemplate`, which nobody complains about. And typical stratum (v1) jobs are refreshed every 30 seconds. https://b10c.me/blog/014-mining-pool-behavior-during-forks/

Reply:
Additionally there's memory management to worry about, if we really start pushing fresh template at every mempool increment, potentially multiple times per second.